File: Stack/M-394. Decode String/M-394. Decode String.py
class Solution:
    @classmethod
    def decodeString(self, s: str) -> str:
        stack = []
        for char in s:
            if char != "]":
                stack.append(char)
                continue

            cur = stack.pop()
            string = ""
            while cur != "[":
                string = cur + string
                cur = stack.pop()

            num = ""
            cur = stack.pop()
            while cur.isnumeric():
                num = cur + num
                cur = stack.pop() if len(stack) > 0 else ""

            if cur:
                stack.append(cur)
            stack.append(string * int(num))
        return "".join(stack)


testData = [
    {
        "s": "3[a]2[bc]",
        "ans": "aaabcbc",
    },
    {
        "s": "3[a2[c]]",
        "ans": "accaccacc",
    },
    {
        "s": "2[abc]3[cd]ef",
        "ans": "abcabccdcdcdef",
    },
    {
        "s": "100[leetcode]",
        "ans": "leetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcode",
    },
]

for data in testData:
    print(data["ans"] == Solution.decodeString(data["s"]))

File: Stack/M-394. Decode String/M-394. Decode String.ts
function decodeString(s: string): string {
  const stack: string[] = [];
  for (const char of s) {
    if (char !== ']') {
      stack.push(char);
      continue;
    }

    //console.log('stack', stack);
    let str = '';
    while (stack[stack.length - 1] !== '[') {
      str = stack.pop() + str;
      //console.log('str', str);
    }

    // pop '['
    stack.pop();

    let num = '';
    while (Number.isInteger(Number(stack[stack.length - 1]))) {
      num = stack.pop() + num;
    }

    stack.push(str.repeat(Number(num)));
  }

  return stack.join('');
}

const testData = [
  {
    s: '3[a]2[bc]',
    ans: 'aaabcbc',
  },
  {
    s: '3[a2[c]]',
    ans: 'accaccacc',
  },
  {
    s: '2[abc]3[cd]ef',
    ans: 'abcabccdcdcdef',
  },
  {
    s: '100[leetcode]',
    ans: 'leetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcode',
  },
];

for (const data of testData) {
  console.log(data['ans'] === decodeString(data['s']));
}

const s = new Set([1, 2, 3]);
for (const c of s) console.log(c);
